hcl2template/addrs/plugin.go

```diff
@@ -0,0 +1,238 @@
+package addrs
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"golang.org/x/net/idna"
+)
+
+// Plugin encapsulates a single plugin type.
+type Plugin struct {
+	Type      string
+	Namespace string
+	Hostname  string
+}
+
+func (p Plugin) RealRelativePath() string {
+	ns := DefaultPluginNamespace
+	if p.Namespace != "" {
+		ns = p.Namespace
+	}
+	return "/" + ns + "/packer-plugin-" + p.Type
+}
+
+func (p Plugin) Parts() []string {
+	return []string{p.Hostname, p.Namespace, p.Type}
+}
+
+func (p Plugin) String() string {
+	return strings.Join(p.Parts(), "/")
+}
+
+// ForDisplay returns a user-friendly FQN string, simplified for readability. If
+// the plugin is using the default hostname, the hostname is omitted.
+func (p *Plugin) ForDisplay() string {
+	parts := []string{}
+	if p.Hostname != DefaultPluginHost {
+		parts = append(parts, p.Hostname)
+	}
+	if p.Namespace != DefaultPluginNamespace {
+		parts = append(parts, p.Namespace)
+	}
+	parts = append(parts, p.Type)
+	return strings.Join(parts, "/")
+}
+
+const DefaultPluginHost = "github.com"
+const DefaultPluginNamespace = "hashicorp"
+
+// ParsePluginPart processes an addrs.Plugin namespace or type string
+// provided by an end-user, producing a normalized version if possible or
+// an error if the string contains invalid characters.
+//
+// A plugin part is processed in the same way as an individual label in a DNS
+// domain name: it is transformed to lowercase per the usual DNS case mapping
+// and normalization rules and may contain only letters, digits, and dashes.
+// Additionally, dashes may not appear at the start or end of the string.
+//
+// These restrictions are intended to allow these names to appear in fussy
+// contexts such as directory/file names on case-insensitive filesystems,
+// repository names on GitHub, etc. We're using the DNS rules in particular,
+// rather than some similar rules defined locally, because the hostname part
+// of an addrs.Plugin is already a hostname and it's ideal to use exactly
+// the same case folding and normalization rules for all of the parts.
+//
```

Comment:
thanks for this docstring. super useful.

Reply:
Kudos to the Terraform team here ! 🙂